### ISSUE TITLE
Workaround for broken cache after bulk delete?

### ISSUE THREAD

**Issue by andreymal** [ponyorm/pony#317]
```python
from pony.orm import *

db = Database()

class Post(db.Entity):
    comments = Set('Comment')
    log = Set('PostEditLog')

class Comment(db.Entity):
    post = Required(Post)
    log = Set('PostEditLog')

class PostEditLog(db.Entity):
    post = Optional(Post)
    comment = Optional(Comment)


db.bind('sqlite', ':memory:', create_db=True)
db.generate_mapping(create_tables=True)
sql_debug(True)

with db_session:
    post = Post()
    c1 = Comment(post=post)
    e1 = PostEditLog(post=post, comment=None)
    e2 = PostEditLog(post=post, comment=c1)

with db_session:
    post = Post.select().first()

    # fill Pony ORM cache
    for c in post.comments:
        c.log.select()[:]

    post.log.select().delete(bulk=True)
    post.delete()
```

This code fails with `Object PostEditLog[2] was updated outside of current transaction`. This is expected behavior, [documentation has note](https://docs.ponyorm.com/api_reference.html?highlight=bulk#Query.delete):

>If an object was loaded into memory, it will not be removed from the db_session() cache on bulk delete.

But I don't want to use normal delete, it's not needed for me. Can I tell Pony that the cache is out of date, something like `pony.orm.db_session.clear_cache()`?

Without restarting of session: single transaction is required.

**Comment by Vitalium** [comment]
I use this workaround:
```
commit()
rollback()
```

Viola!

**Comment by andreymal** [comment]
Single transaction is required for me

**Comment by Vitalium** [comment]
Have you tried to use `flush()`?

**Comment by andreymal** [comment]
Yes, but cache is still broken

**Comment by kozlovsky** [comment]
Cache clearing is very non-trivial task:
1) When someone wants to clear cache. he probably doesn't want to purge *all* objects from the cache, just some of them. For example, in your case you probably don't want to purge `post` object. For user, it is `obvious` which objects he want to keep in the cache, it's an object for which he keep direct references in some variables. But for ORM it is hard to determine which objects are meant to be purged and which should be kept.
2) All objects in cache are interconnected via relationships. In order to purge some objects from the cache, ORM needs to correctly clear links to these objects from other objects. It is a non-trivial operation.

In principle, we can add some partial cache clearing in the future. But it is not clear at this moment how to do it better, and what is correct API.

Right now you have two options:

1) You can use `db_session` with `optimistic=False` option. In this case Pony will not use optimistic checks and will not raise `OptimisticCheckError`. You should use dev version of Pony from GitHub, because I just pushed some fixes in commit 4dc41d29. I tested your example with `db_session(optimistic=False)`, and it works fine.

2) If you exactly know which objects you want to keep in the cache, you can pickle them, start a new `db_session` and unpickle objects to this new `db_session`.

**Comment by andreymal** [comment]
@kozlovsky thanks for the answer! I'll wait for some API for cache control,_ but `optimistic=False` also looks like as suitable workaround for me.
